#5301: Enforce code style described in .clang-format during CI pipelines

esmalTT opened the pull request:
See https://github.com/tenstorrent-metal/tt-metal/issues/5301 for the motivation behind the following fix.

I have added a new static check in CI that will verify all changed files meet the `.clang-format` configuration. The script used by CI also works locally so that developers can easily run the formatter locally, although using the script is not required.  

The new pipeline stage is fast (<1 min) so it can easily be run along with the rest of the static checks.

I also added documentation about the code formatting policy, and instructions on how to use the formatter script. A Slack announcement is also probably a good idea.

From the original issue:

> The main advantage to this solution is it can be gradually rolled out, without disrupting day-to-day development. Over time the repository will become more and more consistent. The downside is that code formatting changes may be noisy for reviewers, and some files that are very rarely (or never!) changed will never become reformatted.

If merged, everyone will need the to install `clang-format` on their machines. For most users it should be a simple `apt-get install`.

Question for reviewers:

- How do you typically test changes like this? Aside from making commits and manually verifying things are working.
- Mark the stage as optional, or hide it away so we can confirm everything is working correctly before accidentally blowing up CI 😅 
- If merged, everyone will need to install `clang-format` on their machines. For most users it should be a simple `apt-get install`.
- There can be slight format differences between major versions of LLVM, which is annoying for users if they happen to run the wrong version.  I've mentioned the correct version in the docs. Do you see any issues with this?
- How do we mark the stage as "Required"?

After feedback I'll clean up the commits.

esmalTT commented:
@TT-billteng Not sure who should review this, but since we talked about it on Slack I'll start with you.

tt-rkim commented on .github/workflows/all-static-checks.yaml:
```diff
@@ -34,6 +34,18 @@ jobs:
+      - name: Install clang-format
+        run: sudo apt-get install -y clang-format
+      - name: Fetch branch base 
+        if: github.ref != 'refs/heads/main'
```
I believe this should be `${{ github.ref !== 'refs/heads/main' }}`, JavaScript style with the braces.

Otherwise this might always evaluate to `true` because it's treated as a string.

tt-rkim replied:
Also, could we add steps for `main` branch to take the `BASE_COMMIT_SHA` as the one directly preceding the current commit?

We have free compute with the github hosted runners. It'd be nice to have this check occur on every main commit as well upon push, even though we should "technically" get it for free on every PR. Just nice for validation and to see the checkmark.

tt-rkim replied:
My bad - it should be `!=` - NOT `!==`

tt-rkim commented on .github/workflows/all-static-checks.yaml:
```diff
@@ -34,6 +34,18 @@ jobs:
+        if: github.ref != 'refs/heads/main'
+        run: git fetch origin $GITHUB_BASE_REF --depth=1
+      - name: Run formatter
+        if: github.ref != 'refs/heads/main'
```
Same deal as above

esmalTT replied:
Fixed

tt-rkim reviewed (changes requested):
Could you also please update the developer dependencies in the README.md on front-page and add `clang-format` as an apt dependency?

tt-rkim commented:
@esmalTT I think one good way to test this would be
- check out a branch with a unique name
- change the base branch from 'main' in the YAML here to the name of new branch and push
- open test PR and see that you get the expected output
- once we're clear, post the link to that and close that test PR

If that makes sense, you could give that a try and report back?

tt-rkim committed:
#5301: Add lint-clang-format and CI pipeline check

tt-rkim committed:
#5308: Various github workflow fixes:
#5308: Condition push event step for clang format only if the event is a push
#5308: Condition pull request events on only if there is a pull request
#5308: Fetch all history for clang format because we're gonna need it for pull requests, and BW is free!

esmalTT commented:
> @esmalTT I think one good way to test this would be
> 
> * check out a branch with a unique name
> * change the base branch from 'main' in the YAML here to the name of new branch and push
> * open test PR and see that you get the expected output
> * once we're clear, post the link to that and close that test PR
> 
> If that makes sense, you could give that a try and report back?

Thanks for taking care of this 😄  I hadn't had a chance to to look at it yet.

tt-rkim reviewed (changes requested):
As per pointed out by @arakhmati @TT-billteng and @cfjchu , we'll need to revisit the call in `lint_clang_format.sh` to ensure that it's only scanning and complaining about the diffed portion of a CPP file

We'll see if this is even possible with clang-format

esmalTT commented:
That makes sense, it should be possible. I’ll see what I can do!
________________________________
From: Raymond Kim ***@***.***>
Sent: Friday, March 1, 2024 2:14:41 PM
To: tenstorrent-metal/tt-metal ***@***.***>
Cc: Evan Smal ***@***.***>; Mention ***@***.***>
Subject: [CAUTION - External Sender] Re: [tenstorrent-metal/tt-metal] #5301: Enforce code style described in .clang-format during CI pipelines (PR #5412)


@tt-rkim requested changes on this pull request.

As per pointed out by @arakhmati<https://www.google.com/url?q=https://github.com/arakhmati&source=gmail-imap&ust=1709925282000000&usg=AOvVaw1X3nyM7x9K1qgtYEbInrKl> @TT-billteng<https://www.google.com/url?q=https://github.com/TT-billteng&source=gmail-imap&ust=1709925282000000&usg=AOvVaw2rSw0OWT-B990DTOm9nK7m> and @cfjchu<https://www.google.com/url?q=https://github.com/cfjchu&source=gmail-imap&ust=1709925282000000&usg=AOvVaw096L4J10soSgyJJQ572Wiw> , we'll need to revisit the call in lint_clang_format.sh to ensure that it's only scanning and complaining about the diffed portion of a CPP file

We'll see if this is even possible with clang-format

—
Reply to this email directly, view it on GitHub<https://www.google.com/url?q=https://github.com/tenstorrent-metal/tt-metal/pull/5412%23pullrequestreview-1911850353&source=gmail-imap&ust=1709925282000000&usg=AOvVaw2Tj5-Yjl0SmGZkBwxIhC85>, or unsubscribe<https://www.google.com/url?q=https://github.com/notifications/unsubscribe-auth/BF6SXPW67C2FV6W52VH2M5TYWDHSDAVCNFSM6AAAAABDKLYDYKVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTSMJRHA2TAMZVGM&source=gmail-imap&ust=1709925282000000&usg=AOvVaw0_36aFyZOEdlnk_P-Qk-B_>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

blozano-tt commented:
> As per pointed out by @arakhmati @TT-billteng and @cfjchu , we'll need to revisit the call in `lint_clang_format.sh` to ensure that it's only scanning and complaining about the diffed portion of a CPP file
> 
> We'll see if this is even possible with clang-format

This wouldn't be so necessary, if there was an agreed upon formatting from developers, and we do one massive re-formatting PR.

tt-rkim commented:
that's right. I think we can do a giant one by making use of https://docs.github.com/en/repositories/working-with-files/using-files/viewing-a-file#ignore-commits-in-the-blame-view

blozano-tt commented:
> that's right. I think we can do a giant one by making use of https://docs.github.com/en/repositories/working-with-files/using-files/viewing-a-file#ignore-commits-in-the-blame-view

After thinking about it more ... a massive formatting change would be higher risk, the formatter could break some functionality, and we wouldn't be able to find which change broke functionality...

blozano-tt commented:
@tt-rkim @esmalTT I just found an open source tool that seems to do what we want:

https://github.com/llvm/llvm-project/blob/main/clang/tools/clang-format/git-clang-format

```
blozano@e03cs01:~/tt-metal$ git-clang-format
The following files would be modified but have unstaged changes:
M	tt_metal/host_api.hpp
Please commit, stage, or stash them first.
blozano@e03cs01:~/tt-metal$ 
blozano@e03cs01:~/tt-metal$ 
blozano@e03cs01:~/tt-metal$ 
blozano@e03cs01:~/tt-metal$ git add tt_metal/
blozano@e03cs01:~/tt-metal$ 
blozano@e03cs01:~/tt-metal$ 
blozano@e03cs01:~/tt-metal$ git-clang-format
changed files:
    tt_metal/host_api.hpp
blozano@e03cs01:~/tt-metal$ 
blozano@e03cs01:~/tt-metal$ 
blozano@e03cs01:~/tt-metal$ git diff 
diff --git a/tt_metal/host_api.hpp b/tt_metal/host_api.hpp
index 71f73aeaa1..dc5ec865f4 100644
--- a/tt_metal/host_api.hpp
+++ b/tt_metal/host_api.hpp
@@ -67,7 +67,7 @@ chip_id_t GetPCIeDeviceID(chip_id_t device_id);
  * |------------|----------------------------|-----------------|-----------------------------------|----------|
  * | device_id  | ID of the device to target| chip_id_t (int) | 0 to (GetNumAvailableDevices - 1) | Yes      |
  * */
-Device *  CreateDevice(
+Device *CreateDevice(
     chip_id_t device_id,
     const uint8_t num_hw_cqs = 1,
     const size_t l1_small_size = DEFAULT_L1_SMALL_SIZE,
```

esmalTT commented:
> @tt-rkim @esmalTT I just found an open source tool that seems to do what we want:
> 
> https://github.com/llvm/llvm-project/blob/main/clang/tools/clang-format/git-clang-format
> 
> ```
> blozano@e03cs01:~/tt-metal$ git-clang-format
> The following files would be modified but have unstaged changes:
> M	tt_metal/host_api.hpp
> Please commit, stage, or stash them first.
> blozano@e03cs01:~/tt-metal$ 
> blozano@e03cs01:~/tt-metal$ 
> blozano@e03cs01:~/tt-metal$ 
> blozano@e03cs01:~/tt-metal$ git add tt_metal/
> blozano@e03cs01:~/tt-metal$ 
> blozano@e03cs01:~/tt-metal$ 
> blozano@e03cs01:~/tt-metal$ git-clang-format
> changed files:
>     tt_metal/host_api.hpp
> blozano@e03cs01:~/tt-metal$ 
> blozano@e03cs01:~/tt-metal$ 
> blozano@e03cs01:~/tt-metal$ git diff 
> diff --git a/tt_metal/host_api.hpp b/tt_metal/host_api.hpp
> index 71f73aeaa1..dc5ec865f4 100644
> --- a/tt_metal/host_api.hpp
> +++ b/tt_metal/host_api.hpp
> @@ -67,7 +67,7 @@ chip_id_t GetPCIeDeviceID(chip_id_t device_id);
>   * |------------|----------------------------|-----------------|-----------------------------------|----------|
>   * | device_id  | ID of the device to target| chip_id_t (int) | 0 to (GetNumAvailableDevices - 1) | Yes      |
>   * */
> -Device *  CreateDevice(
> +Device *CreateDevice(
>      chip_id_t device_id,
>      const uint8_t num_hw_cqs = 1,
>      const size_t l1_small_size = DEFAULT_L1_SMALL_SIZE,
> ```

@blozano-tt Yes this was suggested at some point, I think it is a good option to speed up the linting step. 

There were some concerns that the SortIncludes could break some janky parts of the codebase. If we disabled that option we should be able to run one large format and the hide the commit from the history as suggested by @tt-rkim.

tt-rkim commented:
Transferring some Slack knowledges to here.

@pgkeller and @ayerofieiev-tt acknowledge that we should definitely re-format and have something automated before the team starts shooting up in size. However, doing a one-shot thing for the whole codebase is difficult without disrupting people's work, and possibly introducing regressions.

@blozano-tt has pointed out we can do piecemeal PRs, folder by folder, with buy in from both the TT-NN and tt-metal teams. Then, we slowly add all folders to the required formatter. After all that is done and we all love the new pretty codebase, we can then turn on something like `git-clang-format` or other related tools to only target diffs and lessen the load on our codebase. Some low-hanging fruit could be things like tests and test infrastructure, first.

We should pick this up in priority and establish a folder-by-folder plan if that's the case. No team has to block on the other, as @broskoTT pointed out we can disable formatting by folder, so it can be fully opt-in if people are really concerned that the formatter may cause real compiler diffs.

side-note for @blozano-tt : did you verify that Evan's changes here only work on the diffed lines, or is it all the files in the diff?

tt-rkim commented:
I'm assuminmg we can close this now?

blozano-tt commented:
I am planning to work on manual formatting of the repo before we enable this in CI. Lets leave it open for a bit longer.

blozano-tt commented:
Closing this, as I have implemented/enabled clang-format as a pre-commit hook in CI.
